File: OsmAnd/src/net/osmand/plus/track/helpers/GPXDatabase.java
package net.osmand.plus.track.helpers;

import static net.osmand.IndexConstants.GPX_INDEX_DIR;
import static net.osmand.gpx.GpxParameter.COLOR;
import static net.osmand.gpx.GpxParameter.COLORING_TYPE;
import static net.osmand.gpx.GpxParameter.FILE_CREATION_TIME;
import static net.osmand.gpx.GpxParameter.FILE_DIR;
import static net.osmand.gpx.GpxParameter.FILE_NAME;
import static net.osmand.plus.card.color.ColoringPurpose.TRACK;

import android.util.Pair;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import net.osmand.PlatformUtil;
import net.osmand.gpx.GPXTrackAnalysis;
import net.osmand.gpx.GPXUtilities;
import net.osmand.gpx.GpxParameter;
import net.osmand.plus.OsmandApplication;
import net.osmand.plus.api.SQLiteAPI.SQLiteConnection;
import net.osmand.plus.api.SQLiteAPI.SQLiteCursor;
import net.osmand.plus.routing.ColoringType;
import net.osmand.plus.track.GradientScaleType;
import net.osmand.plus.utils.AndroidDbUtils;
import net.osmand.util.Algorithms;

import org.apache.commons.logging.Log;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class GPXDatabase {

	public static final Log LOG = PlatformUtil.getLog(GPXDatabase.class);

	protected static final int DB_VERSION = 27;
	private static final String DB_NAME = "gpx_database";

	protected static final String GPX_TABLE_NAME = "gpxTable";
	protected static final String GPX_DIR_TABLE_NAME = "gpxDirTable";

	private static final String TMP_NAME_COLUMN_COUNT = "itemsCount";

	public static final long UNKNOWN_TIME_THRESHOLD = 10;

	protected static final String GPX_UPDATE_PARAMETERS_START = "UPDATE " + GPX_TABLE_NAME + " SET ";
	private static final String GPX_FIND_BY_NAME_AND_DIR = " WHERE " + FILE_NAME.getColumnName() + " = ? AND " + FILE_DIR.getColumnName() + " = ?";

	private static final String GPX_MIN_CREATE_DATE = "SELECT " +
			"MIN(" + FILE_CREATION_TIME.getColumnName() + ") " +
			" FROM " + GPX_TABLE_NAME + " WHERE " + FILE_CREATION_TIME.getColumnName() +
			" > " + UNKNOWN_TIME_THRESHOLD;

	private static final String GPX_MAX_COLUMN_VALUE = "SELECT " +
			"MAX(%s) " +
			" FROM " + GPX_TABLE_NAME;

	private static final String CHANGE_NULL_TO_EMPTY_STRING_QUERY_PART = "case when %1$s is null then '' else %1$s end as %1$s";
	private static final String INCLUDE_NON_NULL_COLUMN_CONDITION = " WHERE %1$s NOT NULL AND %1$s <> '' ";
	private static final String GET_ITEM_COUNT_COLLECTION_BASE = "SELECT " + "%s, " + "count (*) as "
			+ TMP_NAME_COLUMN_COUNT + " FROM " + GPX_TABLE_NAME + "%s" + " group by %s" + " ORDER BY %s %s";

	private final OsmandApplication app;

	GPXDatabase(@NonNull OsmandApplication app) {
		this.app = app;
		// init database
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			db.close();
		}
	}

	SQLiteConnection openConnection(boolean readonly) {
		SQLiteConnection conn = app.getSQLiteAPI().getOrCreateDatabase(DB_NAME, readonly);
		if (conn == null) {
			return null;
		}
		if (conn.getVersion() < DB_VERSION) {
			if (readonly) {
				conn.close();
				conn = app.getSQLiteAPI().getOrCreateDatabase(DB_NAME, false);
			}
			if (conn == null) {
				return null;
			}
			int version = conn.getVersion();
			conn.setVersion(DB_VERSION); // not correct version but dangerous for crash loop
			if (version == 0) {
				GpxDbUtils.onCreate(conn);
			} else {
				GpxDbUtils.onUpgrade(this, conn, version, DB_VERSION);
			}
//			conn.setVersion(DB_VERSION); // correct version but dangerous for crash loop
		}
		return conn;
	}

	public boolean updateDataItem(@NonNull DataItem item) {
		File file = item.getFile();
		String tableName = GpxDbUtils.getTableName(file);
		Map<GpxParameter, Object> map = GpxDbUtils.getItemParameters(item);

		return updateGpxParameters(map, tableName, GpxDbUtils.getItemRowsToSearch(app, file));
	}

	private boolean updateGpxParameters(@NonNull Map<GpxParameter, Object> rowsToUpdate,
	                                    @NonNull String tableName, @NonNull Map<String, Object> rowsToSearch) {
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				return updateGpxParameters(db, tableName, rowsToUpdate, rowsToSearch);
			} finally {
				db.close();
			}
		}
		return false;
	}

	private boolean updateGpxParameters(@NonNull SQLiteConnection db, @NonNull String tableName,
	                                    @NonNull Map<GpxParameter, Object> rowsToUpdate,
	                                    @NonNull Map<String, Object> rowsToSearch) {
		Map<String, Object> map = GpxDbUtils.convertGpxParameters(rowsToUpdate);
		Pair<String, Object[]> pair = AndroidDbUtils.createDbUpdateQuery(tableName, map, rowsToSearch);
		//todo (done in ui, move to worker)
		db.execSQL(pair.first, pair.second);
		return true;
	}

	public boolean rename(@NonNull File currentFile, @NonNull File newFile) {
		Map<GpxParameter, Object> map = new LinkedHashMap<>();
		map.put(FILE_NAME, newFile.getName());
		map.put(FILE_DIR, GpxDbUtils.getGpxFileDir(app, newFile));

		String tableName = GpxDbUtils.getTableName(currentFile);
		return updateGpxParameters(map, tableName, GpxDbUtils.getItemRowsToSearch(app, currentFile));
	}

	public boolean remove(@NonNull File file) {
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				String fileName = file.getName();
				String fileDir = GpxDbUtils.getGpxFileDir(app, file);
				String tableName = GpxDbUtils.getTableName(file);
				db.execSQL("DELETE FROM " + tableName + GPX_FIND_BY_NAME_AND_DIR, new Object[] {fileName, fileDir});
			} finally {
				db.close();
			}
			return true;
		}
		return false;
	}

	public boolean add(@NonNull DataItem item) {
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				insertItem(item, db);
			} finally {
				db.close();
			}
			return true;
		}
		return false;
	}

	void insertItem(@NonNull DataItem item, @NonNull SQLiteConnection db) {
		File file = item.getFile();
		String tableName = GpxDbUtils.getTableName(file);
		Map<String, Object> map = GpxDbUtils.convertGpxParameters(GpxDbUtils.getItemParameters(item));
		db.execSQL(AndroidDbUtils.createDbInsertQuery(tableName, map.keySet()), map.values().toArray());
	}

	@NonNull
	private GpxDataItem readGpxDataItem(@NonNull SQLiteCursor query) {
		File file = readItemFile(query);
		GpxDataItem item = new GpxDataItem(app, file);
		GPXTrackAnalysis analysis = new GPXTrackAnalysis();
		processItemParameters(item, query, Arrays.asList(GpxParameter.values()), analysis);
		item.setAnalysis(analysis);

		return item;
	}

	@NonNull
	private File readItemFile(@NonNull SQLiteCursor query) {
		String fileDir = query.getString(query.getColumnIndex(FILE_DIR.getColumnName()));
		String fileName = query.getString(query.getColumnIndex(FILE_NAME.getColumnName()));

		File gpxDir = app.getAppPath(GPX_INDEX_DIR);
		if ((fileName + "/").equalsIgnoreCase(GPX_INDEX_DIR)) {
			return gpxDir;
		}
		fileDir = fileDir.replace(gpxDir.getPath(), "");
		fileDir = fileDir.replace(app.getAppPath(null).getPath(), "");
		File dir = Algorithms.isEmpty(fileDir) ? gpxDir : new File(gpxDir, fileDir);
		return new File(dir, fileName);
	}

	private void processItemParameters(@NonNull DataItem item, @NonNull SQLiteCursor query,
	                                   @NonNull List<GpxParameter> parameters,
	                                   @Nullable GPXTrackAnalysis analysis) {
		for (GpxParameter parameter : parameters) {
			Object value = GpxDbUtils.queryColumnValue(query, parameter);
			if (value == null && !parameter.isAppearanceParameter()) {
				value = parameter.getDefaultValue();
			}
			if (parameter.isAnalysisParameter()) {
				if (analysis != null) {
					analysis.setGpxParameter(parameter, value);
				}
			} else {
				if (parameter == COLOR) {
					value = GPXUtilities.parseColor((String) value);
				} else if (parameter == COLORING_TYPE) {
					String type = (String) value;
					ColoringType coloringType = ColoringType.valueOf(TRACK, type);
					GradientScaleType scaleType = GradientScaleType.getGradientTypeByName(type);
					if (coloringType == null && scaleType != null) {
						coloringType = ColoringType.valueOf(scaleType);
						value = coloringType == null ? null : coloringType.getName(null);
					}
				}
				item.setParameter(parameter, value);
			}
		}
	}

	@NonNull
	private GpxDirItem readGpxDirItem(@NonNull SQLiteCursor query) {
		File file = readItemFile(query);
		GpxDirItem item = new GpxDirItem(app, file);
		processItemParameters(item, query, GpxParameter.getGpxDirParameters(), null);

		return item;
	}

	public long getTracksMinCreateDate() {
		long minDate = -1;
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				SQLiteCursor query = db.rawQuery(GPX_MIN_CREATE_DATE, null);
				if (query != null) {
					try {
						if (query.moveToFirst()) {
							minDate = query.getLong(0);
						}
					} finally {
						query.close();
					}
				}
			} finally {
				db.close();
			}
		}
		return minDate;
	}

	public String getColumnMaxValue(@NonNull GpxParameter parameter) {
		String maxValue = "";
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				String queryString = String.format(GPX_MAX_COLUMN_VALUE, parameter.getColumnName());
				SQLiteCursor query = db.rawQuery(queryString, null);
				if (query != null) {
					try {
						if (query.moveToFirst()) {
							maxValue = query.getString(0);
						}
					} finally {
						query.close();
					}
				}
			} finally {
				db.close();
			}
		}
		return maxValue;
	}

	@NonNull
	public List<Pair<String, Integer>> getStringIntItemsCollection(@NonNull String columnName,
	                                                               boolean includeEmptyValues,
	                                                               boolean sortByName,
	                                                               boolean sortDescending) {
		String column1 = includeEmptyValues ? String.format(CHANGE_NULL_TO_EMPTY_STRING_QUERY_PART, columnName) : columnName;
		String includeEmptyValuesPart = includeEmptyValues ? "" : String.format(INCLUDE_NON_NULL_COLUMN_CONDITION, columnName);
		String orderBy = sortByName ? columnName : TMP_NAME_COLUMN_COUNT;
		String sortDirection = sortDescending ? "DESC" : "ASC";
		String query = String.format(GET_ITEM_COUNT_COLLECTION_BASE, column1, includeEmptyValuesPart, columnName, orderBy, sortDirection);
		return getStringIntItemsCollection(query);
	}

	@NonNull
	private List<Pair<String, Integer>> getStringIntItemsCollection(@NonNull String dataQuery) {
		List<Pair<String, Integer>> folderCollection = new ArrayList<>();
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				SQLiteCursor query = db.rawQuery(dataQuery, null);
				if (query != null) {
					try {
						if (query.moveToFirst()) {
							do {
								folderCollection.add(new Pair<>(query.getString(0), query.getInt(1)));
							} while (query.moveToNext());
						}
					} finally {
						query.close();
					}
				}
			} finally {
				db.close();
			}
		}
		return folderCollection;
	}

	@NonNull
	public List<GpxDataItem> getGpxDataItems() {
		Set<GpxDataItem> items = new HashSet<>();
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				SQLiteCursor query = db.rawQuery(GpxDbUtils.getSelectGpxQuery(), null);
				if (query != null) {
					try {
						if (query.moveToFirst()) {
							do {
								items.add(readGpxDataItem(query));
							} while (query.moveToNext());
						}
					} finally {
						query.close();
					}
				}
			} finally {
				db.close();
			}
		}
		return new ArrayList<>(items);
	}

	@NonNull
	public List<GpxDirItem> getGpxDirItems() {
		Set<GpxDirItem> items = new HashSet<>();
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				SQLiteCursor query = db.rawQuery(GpxDbUtils.getSelectGpxDirQuery(), null);
				if (query != null) {
					try {
						if (query.moveToFirst()) {
							do {
								items.add(readGpxDirItem(query));
							} while (query.moveToNext());
						}
					} finally {
						query.close();
					}
				}
			} finally {
				db.close();
			}
		}
		return new ArrayList<>(items);
	}

	@Nullable
	public GpxDataItem getGpxDataItem(@NonNull File file) {
		if (GpxUiHelper.isGpxFile(file)) {
			return (GpxDataItem) getDataItem(file);
		}
		return null;
	}

	@Nullable
	public GpxDirItem getGpxDirItem(@NonNull File file) {
		if (file.isDirectory()) {
			return (GpxDirItem) getDataItem(file);
		}
		return null;
	}

	@Nullable
	private DataItem getDataItem(@NonNull File file) {
		DataItem result = null;
		SQLiteConnection db = openConnection(false);
		if (db != null) {
			try {
				result = getDataItem(file, db);
			} finally {
				db.close();
			}
		}
		return result;
	}

	@Nullable
	public DataItem getDataItem(@NonNull File file, @NonNull SQLiteConnection db) {
		String name = file.getName();
		String dir = GpxDbUtils.getGpxFileDir(app, file);
		boolean gpxFile = GpxUiHelper.isGpxFile(file);

		String selectQuery = gpxFile ? GpxDbUtils.getSelectGpxQuery() : GpxDbUtils.getSelectGpxDirQuery();
		SQLiteCursor query = db.rawQuery(selectQuery + GPX_FIND_BY_NAME_AND_DIR, new String[] {name, dir});
		if (query != null) {
			try {
				if (query.moveToFirst()) {
					return gpxFile ? readGpxDataItem(query) : readGpxDirItem(query);
				}
			} finally {
				query.close();
			}
		}
		return null;
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/gpx/GpxDatabase.kt
package net.osmand.shared.gpx

import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import net.osmand.shared.api.SQLiteAPI.*
import net.osmand.shared.data.StringIntPair
import net.osmand.shared.extensions.currentTimeMillis
import net.osmand.shared.extensions.format
import net.osmand.shared.gpx.GpxParameter.*
import net.osmand.shared.io.KFile
import net.osmand.shared.routing.ColoringType
import net.osmand.shared.util.DbUtils
import net.osmand.shared.util.LoggerFactory
import net.osmand.shared.util.PlatformUtil

class GpxDatabase {

	companion object {
		val log = LoggerFactory.getLogger("GpxDatabase")

		const val DB_VERSION = 32
		const val DB_NAME = "gpx_database"
		const val GPX_TABLE_NAME = "gpxTable"
		const val GPX_DIR_TABLE_NAME = "gpxDirTable"
		const val TMP_NAME_COLUMN_COUNT = "itemsCount"
		const val UNKNOWN_TIME_THRESHOLD = 10L
		val GPX_UPDATE_PARAMETERS_START = "UPDATE $GPX_TABLE_NAME SET "
		val GPX_FIND_BY_NAME_AND_DIR = " WHERE ${FILE_NAME.columnName} = ? AND ${FILE_DIR.columnName} = ?"
		val GPX_NAME_AND_DIR = "${FILE_NAME.columnName} = ? AND ${FILE_DIR.columnName} = ?"
		val GPX_MIN_CREATE_DATE =
			"SELECT MIN(${FILE_CREATION_TIME.columnName}) FROM $GPX_TABLE_NAME WHERE ${FILE_CREATION_TIME.columnName} > $UNKNOWN_TIME_THRESHOLD"
		val GPX_MAX_COLUMN_VALUE = "SELECT MAX(%s) FROM $GPX_TABLE_NAME"
		val CHANGE_NULL_TO_EMPTY_STRING_QUERY_PART =
			"case when %1\$s is null then '' else %1\$s end as %1\$s"
		val CHANGE_NULL_TO_EMPTY_GROUP_CONDITION_STRING_QUERY_PART =
			"case when %1\$s is null then '' else %1\$s end"
		val INCLUDE_NON_NULL_COLUMN_CONDITION = " WHERE %1\$s NOT NULL AND %1\$s <> '' "
		val GET_ITEM_COUNT_COLLECTION_BASE =
			"SELECT %s, count (*) as $TMP_NAME_COLUMN_COUNT FROM $GPX_TABLE_NAME%s group by %s ORDER BY %s %s"

		val BATCH_SIZE = 100
	}

	init {
		val db = openConnection(false)
		db?.close()
	}

	fun openConnection(readonly: Boolean): SQLiteConnection? {
		var conn = PlatformUtil.getSQLiteAPI().getOrCreateDatabase(DB_NAME, readonly)
		if (conn == null) return null

		if (conn.getVersion() < DB_VERSION) {
			if (readonly) {
				conn.close()
				conn = PlatformUtil.getSQLiteAPI().getOrCreateDatabase(DB_NAME, false)
			}
			if (conn == null) return null
			val version = conn.getVersion()
			conn.setVersion(DB_VERSION) // not correct version but dangerous for crash loop
			if (version == 0) {
				GpxDbUtils.onCreate(conn)
			} else {
				GpxDbUtils.onUpgrade(this, conn, version, DB_VERSION)
			}
			//conn.setVersion(DB_VERSION); // correct version but dangerous for crash loop
		}
		return conn
	}

	fun updateDataItem(item: DataItem): Boolean {
		return updateGpxParameters(item, GpxDbUtils.getItemParameters(item))
	}

	fun updateDataItemParameter(item: DataItem, gpxParameter: GpxParameter, value: Any?): Boolean {
		return updateGpxParameters(item, linkedMapOf(gpxParameter to value))
	}

	private fun updateGpxParameters(item: DataItem, map: Map<GpxParameter, Any?>): Boolean {
		val file = item.file
		val tableName = GpxDbUtils.getTableName(file)
		val success = updateGpxParameters(map, tableName, GpxDbUtils.getItemRowsToSearch(file))

		if (success && item is GpxDataItem) {
			updateAppearanceTimestamp(item)
		}
		return success
	}

	private fun updateGpxParameters(
		rowsToUpdate: Map<GpxParameter, Any?>,
		tableName: String,
		rowsToSearch: Map<String, Any?>
	): Boolean {
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			return db?.let { updateGpxParameters(it, tableName, rowsToUpdate, rowsToSearch) } ?: false
		} finally {
			db?.close()
		}
	}

	private fun updateGpxParameters(
		db: SQLiteConnection,
		tableName: String,
		rowsToUpdate: Map<GpxParameter, Any?>,
		rowsToSearch: Map<String, Any?>
	): Boolean {
		val map = GpxDbUtils.convertGpxParameters(rowsToUpdate)
		val pair = DbUtils.createDbUpdateQuery(tableName, map, rowsToSearch)
		db.execSQL(pair.first, pair.second)
		return true
	}

	fun rename(currentFile: KFile, newFile: KFile): Boolean {
		val map = linkedMapOf(
			FILE_NAME to newFile.name(),
			FILE_DIR to GpxDbUtils.getGpxFileDir(newFile)
		)
		val tableName = GpxDbUtils.getTableName(currentFile)
		return updateGpxParameters(map, tableName, GpxDbUtils.getItemRowsToSearch(currentFile))
	}

	fun remove(file: KFile): Boolean {
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			db?.let {
				val fileName = file.name()
				val fileDir = GpxDbUtils.getGpxFileDir(file)
				val tableName = GpxDbUtils.getTableName(file)
				db.execSQL(
					"DELETE FROM $tableName $GPX_FIND_BY_NAME_AND_DIR",
					arrayOf(fileName, fileDir)
				)
				return true
			}
			return false
		} finally {
			db?.close()
		}
	}

	fun remove(files: Collection<KFile>): Boolean {
		if (files.isEmpty()) return false

		val time = currentTimeMillis()
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			db?.let {
				val fileDeletionMap = files.groupBy { GpxDbUtils.getTableName(it) }
				fileDeletionMap.forEach { (tableName, files) ->
					files.chunked(BATCH_SIZE).forEach { batch ->
						val deleteConditions = batch.joinToString(separator = " OR ") { "($GPX_NAME_AND_DIR)" }
						val args: Array<Any?> = batch.flatMap { listOf(it.name(), GpxDbUtils.getGpxFileDir(it)) }.toTypedArray()
						db.execSQL("DELETE FROM $tableName WHERE $deleteConditions", args)
					}
				}
				log.info("Remove gpx files from db count=${files.size} in ${currentTimeMillis() - time} ms")
				return true
			}
			return false
		} finally {
			db?.close()
		}
	}

	fun add(item: DataItem): Boolean {
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			return db?.let {
				insertItem(item, it)
				true
			} ?: false
		} finally {
			db?.close()
		}
	}

	fun insertItem(item: DataItem, db: SQLiteConnection) {
		val file = item.file
		val tableName = GpxDbUtils.getTableName(file)
		val map = GpxDbUtils.convertGpxParameters(GpxDbUtils.getItemParameters(item))
		val query = DbUtils.createDbInsertQuery(tableName, map.keys)
		db.execSQL(query, map.values.toTypedArray())
	}

	private fun readGpxDataItem(query: SQLiteCursor): GpxDataItem {
		val file = readItemFile(query)
		val item = GpxDataItem(file)
		val analysis = GpxTrackAnalysis()
		processItemParameters(item, query, entries, analysis)
		item.setAnalysis(analysis)
		return item
	}

	private fun readItemFile(query: SQLiteCursor): KFile {
		var fileDir: String = query.getString(query.getColumnIndex(FILE_DIR.columnName))
		val fileName = query.getString(query.getColumnIndex(FILE_NAME.columnName))

		val appDir = PlatformUtil.getOsmAndContext().getAppDir()
		val gpxDir = PlatformUtil.getOsmAndContext().getGpxDir()
		if (fileName == gpxDir.name()) {
			return gpxDir
		}
		fileDir = fileDir.replace(gpxDir.toString(), "")
		fileDir = fileDir.replace(appDir.toString(), "")
		val dir = if (fileDir.isEmpty()) gpxDir else KFile(gpxDir, fileDir)
		return KFile(dir, fileName)
	}

	private fun processItemParameters(
		item: DataItem,
		query: SQLiteCursor,
		parameters: List<GpxParameter>,
		analysis: GpxTrackAnalysis?
	) {
		for (parameter in parameters) {
			var value = GpxDbUtils.queryColumnValue(query, parameter)
			if (value == null && !parameter.isAppearanceParameter()) {
				value = parameter.defaultValue
			}
			if (parameter.analysisParameter) {
				analysis?.setGpxParameter(parameter, value)
			} else {
				if (parameter == COLOR) {
					value = GpxUtilities.parseColor(value as String?)
				} else if (parameter == COLORING_TYPE) {
					val type = value as String?
					var coloringType = ColoringType.valueOf(ColoringPurpose.TRACK, type)
					val scaleType = GradientScaleType.getGradientTypeByName(type)
					if (coloringType == null && scaleType != null) {
						coloringType = ColoringType.valueOf(scaleType)
						value = coloringType?.getName(null)
					}
				}
				item.setParameter(parameter, value)
			}
		}
	}

	private fun readGpxDirItem(query: SQLiteCursor): GpxDirItem {
		val file = readItemFile(query)
		val item = GpxDirItem(file)
		processItemParameters(item, query, GpxParameter.getGpxDirParameters(), null)
		return item
	}

	fun getTracksMinCreateDate(): Long {
		var minDate = -1L
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			db?.let {
				var query: SQLiteCursor? = null
				try {
					query = db.rawQuery(GPX_MIN_CREATE_DATE, null)
					if (query != null && query.moveToFirst()) {
						minDate = query.getLong(0)
					}
				} finally {
					query?.close()
				}
			}
		} finally {
			db?.close()
		}
		return minDate
	}

	fun getColumnMaxValue(parameter: GpxParameter): String {
		var maxValue = ""
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			if (db != null) {
				val queryString = GPX_MAX_COLUMN_VALUE.format(parameter.columnName)
				var query: SQLiteCursor? = null
				try {
					query = db.rawQuery(queryString, null)
					if (query != null && query.moveToFirst()) {
						maxValue = query.getString(0)
					}
				}
				catch (error: Throwable){
					//Cursor.getString may produce exception when the column value is null or the column type is not a string type
				}
				finally {
					query?.close()
				}
			}
		} finally {
			db?.close()
		}
		return maxValue
	}

	fun getStringIntItemsCollection(
		columnName: String,
		includeEmptyValues: Boolean,
		sortByName: Boolean,
		sortDescending: Boolean
	): List<StringIntPair> {
		val column1 =
			if (includeEmptyValues) CHANGE_NULL_TO_EMPTY_STRING_QUERY_PART.format(columnName) else columnName
		val group =
			if (includeEmptyValues) CHANGE_NULL_TO_EMPTY_GROUP_CONDITION_STRING_QUERY_PART.format(columnName) else columnName
		val includeEmptyValuesPart =
			if (includeEmptyValues) "" else INCLUDE_NON_NULL_COLUMN_CONDITION.format(columnName)
		val orderBy = if (sortByName) columnName else TMP_NAME_COLUMN_COUNT
		val sortDirection = if (sortDescending) "DESC" else "ASC"
		val query = GET_ITEM_COUNT_COLLECTION_BASE.format(
			column1,
			includeEmptyValuesPart,
			group,
			orderBy,
			sortDirection
		)
		return getStringIntItemsCollection(query)
	}

	private fun getStringIntItemsCollection(dataQuery: String): List<StringIntPair> {
		val folderCollection = mutableListOf<StringIntPair>()
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			db?.let {
				var query: SQLiteCursor? = null
				try {
					query = db.rawQuery(dataQuery, null)
					if (query != null && query.moveToFirst()) {
						do {
							folderCollection.add(StringIntPair(query.getString(0), query.getInt(1)))
						} while (query.moveToNext())
					}
				} finally {
					query?.close()
				}
			}
		} finally {
			db?.close()
		}
		return folderCollection
	}

	fun getGpxDataItemsSync(): List<GpxDataItem> {
		val items = mutableSetOf<GpxDataItem>()
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			db?.let {
				var query: SQLiteCursor? = null
				try {
					query = db.rawQuery(GpxDbUtils.getSelectGpxQuery(), null)
					if (query != null && query.moveToFirst()) {
						do {
							items.add(readGpxDataItem(query))
						} while (query.moveToNext())
					}
				} finally {
					query?.close()
				}
			}
		} finally {
			db?.close()
		}
		return items.toList()
	}

	fun getGpxDataItemsBlocking(): List<GpxDataItem> = runBlocking { getGpxDataItems() }

	suspend fun getGpxDataItems(): List<GpxDataItem> = coroutineScope {
		val items = mutableListOf<GpxDataItem>()
		val deferredResults = mutableListOf<Deferred<List<GpxDataItem>>>()
		var offset = 0
		val itemsCount = getGpxDirItemsCount()
		while (offset < itemsCount) {
			val currentOffset = offset
			val deferredBatch = async(Dispatchers.IO) {
				var db: SQLiteConnection? = null
				try {
					db = openConnection(true)
					if (db != null) {
						fetchBatchData(db, currentOffset, BATCH_SIZE)
					} else {
						emptyList()
					}
				} finally {
					db?.close()
				}
			}
			deferredResults.add(deferredBatch)
			offset += BATCH_SIZE
		}

		deferredResults.awaitAll().forEach { batchItems ->
			items.addAll(batchItems)
		}
		return@coroutineScope items.toList()
	}

	private fun getGpxDirItemsCount(): Int {
		var res = 0
		var db: SQLiteConnection? = null
		try {
			db = openConnection(true)
			db?.let {
				var query: SQLiteCursor? = null
				try {
					query = db.rawQuery("SELECT COUNT(*) FROM $GPX_TABLE_NAME", null)
					if (query != null && query.moveToFirst()) {
						res = query.getInt(0)
					}
				} finally {
					query?.close()
				}
			}
		} finally {
			db?.close()
		}
		return res
	}

	private fun fetchBatchData(db: SQLiteConnection, offset: Int, batchSize: Int): List<GpxDataItem> {
		val time = currentTimeMillis()
		val batchItems = mutableListOf<GpxDataItem>()
		var query: SQLiteCursor? = null
		try {
			val paginatedQuery = "${GpxDbUtils.getSelectGpxQuery()} ORDER BY ${FILE_NAME.columnName} LIMIT $batchSize OFFSET $offset"
			query = db.rawQuery(paginatedQuery, null)
			if (query != null && query.moveToFirst()) {
				do {
					batchItems.add(readGpxDataItem(query))
				} while (query.moveToNext())
			}
		} finally {
			query?.close()
		}
		log.info("loadGpxItems fetchBatchData offset=$offset batchItems=${batchItems.size} in ${currentTimeMillis() - time} ms")
		return batchItems
	}

	fun getGpxDirItems(): List<GpxDirItem> {
		val items = mutableSetOf<GpxDirItem>()
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			db?.let {
				var query: SQLiteCursor? = null
				try {
					query = db.rawQuery(GpxDbUtils.getSelectGpxDirQuery(), null)
					if (query != null && query.moveToFirst()) {
						do {
							items.add(readGpxDirItem(query))
						} while (query.moveToNext())
					}
				} finally {
					query?.close()
				}
			}

		} finally {
			db?.close()
		}
		return items.toList()
	}

	fun getGpxDataItem(file: KFile): GpxDataItem? {
		return if (GpxDbUtils.isGpxFile(file)) getDataItem(file) as? GpxDataItem else null
	}

	fun getGpxDirItem(file: KFile): GpxDirItem? {
		return if (file.isDirectory()) getDataItem(file) as? GpxDirItem else null
	}

	private fun getDataItem(file: KFile): DataItem? {
		var db: SQLiteConnection? = null
		try {
			db = openConnection(false)
			db?.let { return getDataItem(file, db) } ?: return null
		} finally {
			db?.close()
		}
	}

	fun getDataItem(file: KFile, db: SQLiteConnection): DataItem? {
		val name = file.name()
		val dir = GpxDbUtils.getGpxFileDir(file)
		val gpxFile = GpxDbUtils.isGpxFile(file)
		val selectQuery =
			if (gpxFile) GpxDbUtils.getSelectGpxQuery() else GpxDbUtils.getSelectGpxDirQuery()
		var query: SQLiteCursor? = null
		try {
			query = db.rawQuery("$selectQuery $GPX_FIND_BY_NAME_AND_DIR", arrayOf(name, dir))
			if (query != null && query.moveToFirst()) {
				return if (gpxFile) readGpxDataItem(query) else readGpxDirItem(query)
			}
		} finally {
			query?.close()
		}
		return null
	}

	fun isDataItemExists(file: KFile, db: SQLiteConnection): Boolean {
		val name = file.name()
		val dir = GpxDbUtils.getGpxFileDir(file)
		val gpxFile = GpxDbUtils.isGpxFile(file)
		val selectQuery =
			if (gpxFile) GpxDbUtils.getSelectGpxQuery(FILE_NAME) else GpxDbUtils.getSelectGpxDirQuery(FILE_NAME)
		var query: SQLiteCursor? = null
		try {
			query = db.rawQuery("$selectQuery $GPX_FIND_BY_NAME_AND_DIR", arrayOf(name, dir))
			if (query != null && query.moveToFirst()) {
				return true
			}
		} finally {
			query?.close()
		}
		return false
	}

	private fun updateAppearanceTimestamp(item: GpxDataItem) {
		var db: SQLiteConnection? = null
		try {
			db = openConnection(true)
			db?.let {
				var cursor: SQLiteCursor? = null
				try {
					val fileName = item.file.name()
					val fileDir = GpxDbUtils.getGpxFileDir(item.file)
					val query = "${GpxDbUtils.getSelectGpxQuery(APPEARANCE_LAST_MODIFIED_TIME)} $GPX_FIND_BY_NAME_AND_DIR"

					cursor = it.rawQuery(query, arrayOf(fileName, fileDir))
					if (cursor != null && cursor.moveToFirst()) {
						val timestamp = cursor.getLong(0)
						item.setParameter(APPEARANCE_LAST_MODIFIED_TIME, timestamp)
					}
				} finally {
					cursor?.close()
				}
			}
		} finally {
			db?.close()
		}
	}
}